[RFC] Generic SPI Flash driver

fabled opened the pull request:
Initial implementation of generic SPI flash driver supporting SFDP enabled chips (all new ones). It is written in a way that addition support for faster modes is relatively simple. SFDP also provides sector/block layout so generic support for that can be implemented.

Sponsored by @CryptaLabs

fabled committed:
driver: spi: add a generic SPI flash driver

This drivers supports chips with Serial Flash Discoverable
Parameters (SFDP) information.

Tested with CEC1702 QMSPI host controller (PR #5383) connected
to SST26VF016B flash on a SecureIoT1702 board.

Signed-off-by: Timo Teräs <timo.teras@iki.fi>

superna9999 commented:
Pretty cool driver, usage of the SFDP table is a good feature !

With which flash device and HW have you tested it ?

superna9999 commented on drivers/flash/spi_flash.c:
```diff
@@ -0,0 +1,523 @@
+	return r;
+}
+
+
```
Remove blank line

superna9999 commented on drivers/flash/spi_flash.c:
```diff
@@ -0,0 +1,523 @@
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	/* FIXME: Support chip erase, and large block erase */
```
How are you sure all chips support 4k erase ?

fabled replied:
I think it's sort of requirement. In any case the presence of this opcode is tested above.

superna9999 commented on drivers/flash/spi_flash.c:
```diff
@@ -0,0 +1,523 @@
+	buf[0] = cmd_id;
+	if (spi_transceive(&data->config, &cmd, 1, &cmd, 1) != 0) {
+		return -EIO;
+	}
```
Add blank line

fabled commented:
Review comments addressed, test hardware documented in commit message.

pfalcon commented:
@fabled: Build for your patch still fails, you would need to fix that to proceed. Thanks.

pfalcon commented:
recheck

tbursztyka commented on drivers/flash/spi_flash.c:
```diff
@@ -0,0 +1,523 @@
+	}
+
+	nph = min(ARRAY_SIZE(phdr), sfdp.nph + 1UL);
+	r = spi_flash_read_sfdp(dev, sizeof sfdp, phdr, sizeof(struct sfdp_parameter_header[nph]));
```
beware of the 80 chars limit

tbursztyka commented on drivers/flash/spi_flash.c:
```diff
@@ -0,0 +1,523 @@
+{
+	struct spi_flash_data *const data = dev->driver_data;
+	uint8_t buf[4] = { CMD_IDENTIFY };
+	struct spi_buf bufs = { .buf = &buf, sizeof buf };
```
always 1 line per attribute assignement. (you are actually doing it above, line 160, properly). 

Apply that everywhere.

tbursztyka commented on drivers/flash/spi_flash.c:
```diff
@@ -0,0 +1,523 @@
+		sfdp.major_ver, sfdp.minor_ver,
+		(data->flash_size + 1) >> 10,
+		data->page_size);
+	if (!data->bfp.density) {
```
sounds like you should be moving this test before line 258.

tbursztyka commented on drivers/flash/spi_flash.c:
```diff
@@ -0,0 +1,523 @@
+		.frequency = CONFIG_SPI_FLASH_SPI_FREQ_0,
+		.operation = SPI_WORD_SET(8),
+	};
+
```
You will need to add support for gpio based CS. See what I did, for instance, in https://github.com/zephyrproject-rtos/zephyr/pull/5217/commits/3caa3ba293dbe5954c84a394d906865cf9d609c6 line 746.

tbursztyka commented on drivers/flash/spi_flash.c:
```diff
@@ -0,0 +1,523 @@
+	if (r != 0) {
+		return r;
+	}
+	if (sfdp.signature != 0x50444653) {
```
magic number? Make it into a macro (and document it)

tbursztyka commented on drivers/spi/spi_context.h:
```diff
@@ -208,8 +208,10 @@ void spi_context_update_tx(struct spi_context *ctx, u8_t dfs, u32_t len)
+		do {
+			ctx->current_tx++;
+			ctx->tx_count--;
+		} while (ctx->tx_count && ctx->current_tx->len == 0);
```
what you are doing here does not exactly match your commit message. You'll need to change your commit message.

What is done here is avoiding to jump on a bogus buffer entry where pointer is set, but with a length of 0.
Fact that it affects spi peripherals is indirect.

fabled commented:
Addressed most stylistic issues. Separation of .h file and GPIO based CS support still pending. It also seems the github test failure was due to stylistic issues which should get now fixed (I installed checkpatch to git hooks only afterwards, and only now realized it needs to go to post-rewrite hook to be ran in commit --amend).

henrikbrixandersen commented on drivers/flash/spi_flash.c:
```diff
@@ -0,0 +1,523 @@
+#define STATUS_BUSY		0x01
+
+#define MAX_SFDP_HEADERS	4
+
```
Perhaps it would be wise to name it e.g. `drivers/flash/spi_flash_sfdp.[ch]` since we also have other SPI flash drivers in the tree (currently, we have one other SPI flash driver)?

fabled commented:
Addressed file name review comment, renamed also Kconfig options similarly. Rebased against current master.

fabled committed:
spi_context: ignore spi_buf elements with zero .len

The vector can contain elements with valid pointer but zero length
for reason or another. This commit allows having them, as it may
simplify writing spi perhipheral drivers.

Signed-off-by: Timo Teräs <timo.teras@iki.fi>

nashif commented:
stale PR, closing.
